platforms/software/dependency-management/src/main/java/org/gradle/internal/component/model/DefaultAttributeMatcher.java

```diff
@@ -62,28 +62,75 @@ public AttributeSelectionSchema getSelectionSchema() {
     }
 
     @Override
-    public boolean isMatching(AttributeContainerInternal candidate, AttributeContainerInternal requested) {
+    public boolean isMatching(ImmutableAttributes candidate, ImmutableAttributes requested) {
         if (requested.isEmpty() || candidate.isEmpty()) {
             return true;
         }
 
-        ImmutableAttributes requestedAttributes = requested.asImmutable();
-        ImmutableAttributes candidateAttributes = candidate.asImmutable();
+        for (Attribute<?> attribute : requested.keySet()) {
+            AttributeValue<?> requestedAttributeValue = requested.findEntry(attribute);
+            AttributeValue<?> candidateAttributeValue = candidate.findEntry(attribute.getName());
+
+            Attribute<?> typedAttribute = getTypedAttribute(attribute);
+            if (candidateAttributeValue.isPresent()) {
+                Object requestedValue = requestedAttributeValue.coerce(typedAttribute);
+                Object candidateValue = candidateAttributeValue.coerce(typedAttribute);
+
+                boolean match = schema.matchValue(typedAttribute, requestedValue, candidateValue);
 
-        for (Attribute<?> attribute : requestedAttributes.keySet()) {
-            AttributeValue<?> requestedValue = requestedAttributes.findEntry(attribute);
-            AttributeValue<?> candidateValue = candidateAttributes.findEntry(attribute.getName());
-            if (candidateValue.isPresent()) {
-                Object coercedValue = candidateValue.coerce(attribute);
-                boolean match = schema.matchValue(attribute, requestedValue.get(), coercedValue);
                 if (!match) {
                     return false;
                 }
             }
         }
+
         return true;
     }
 
+    @Override
+    public boolean weaklyMatches(ImmutableAttributes first, ImmutableAttributes second) {
```

Comment:
🤔 Can `isMatching` and `weaklyMatches` both forward to a private method that takes a `BiPredicate` for the match logic?

Reply:
These are the same algoritm except one checks one direction and the other checks both directions, so we could theoretically combine these. 

However, the "not common" parts of these two methods take _three_ parameters, instead of two. That means we would need a `TriPredicate`, and would need to define it here. Is that worth it?